### PR TITLE
Added _mac attribute initialization

### LINKED ISSUE #3: Starter Task - Add MAC attribute
#  Add MAC attribute to MPCTensor class

In Python, objects have access to a list of attributes contained in their `self` variable. We can add attributes to `self` by simple assignment:

`self._mac = 0`

In this task, we would like to add a `_mac` attribute when initializing an instance of the `MPCTensor` class. The initialization code can be found [here](https://github.com/MLH-Fellowship/CrypTen/blob/master/crypten/mpc/mpc.py#L110-L147). For now, we don't care what value this attribute takes, so we can set it to an arbitrary value (zero is fine). 

(Note: The underscore in the name `_mac` is meant to denote that this attribute should not be accessed by external users. While nothing prevents an external user from accessing and modifying the `_mac` attribute, the underscore in the name tells the user that they cannot rely on that attribute functioning in an expected way, and that we may remove or change the function of the `_mac` attribute at any time)

Even though we don't care about its value, we should ensure that the `_mac` attribute propagates to any new MPCTensor object that is created from the initial MPCTensor. That is, when computing `y = x + x` should populate a `_mac` attribute in `y`. Hint: To start, try implementing this for the [clone function](https://github.com/MLH-Fellowship/CrypTen/blob/master/crypten/mpc/mpc.py#L164-L184).

## Unit testing

In order to ensure this task is completed correctly, we should add unit tests to [CrypTen's testing harness](https://github.com/MLH-Fellowship/CrypTen/tree/master/test) that assert correctness in a wide variety of scenarios. Because our existing unit tests cover all functions we care about, we can rely on the coverage of these tests. These can be found in [test_mpc.py](https://github.com/MLH-Fellowship/CrypTen/blob/master/test/test_mpc.py). 

Because all of our existing unit tests check correctness using the [_check function](https://github.com/MLH-Fellowship/CrypTen/blob/master/test/test_mpc.py#L34-L60), we should be able to add testing for `_mac` by adding the following two lines to our `_check` function:

```
self.assertTrue(hasattr(encrypted_tensor, "_mac"), "_mac attribute failed to propagate")
self.assertEqual(encrypted_tensor._mac, 0, "_mac attribute has incorrect value")
```

To run unit tests, you can use Python's `unittest` framework. To run all tests in test_mpc.py from the command-line, run:

`python -m unittest test/test_mpc.py`

To run a single test from the command-line, run:

`python -m unittest test/test_mpc.py -k test.test_mpc.TestTFP.<test_name>`

Note:
Some of the unit tests are currently failing. This is atypical and bad practice. In general, PR's should not land if they break any unit tests. However, due to some changes in CrypTen's dependencies, we have a few spots that are broken and should be fixed by next week. To start, just run the following tests which I've verified should work as intended:

test_copy_clone
test_index_select
test_view_flatten
test_arithmetic
test_matmul
test_get_set

## Debugging

It is common practice when debugging in Python to use the Python debugger tool (`pdb`). `pdb` gives the user access to a command prompt during execution of their code. This allows the user to step through their code line-by-line and examine the state of all variables within a given scope. Users can enter this debugging state by running the code `pdb.set_trace()`.

When using CrypTen, we are often operating in a regime where multiple processes are running simultaneously. This can cause catastrophic failures in pdb (trust me, don't do this). There are two ways to counteract this when debugging for this task:

1. Multiprocessing can be turned off in the unit tests by setting the `world_size` property in [multiprocess_test_case.py](https://github.com/MLH-Fellowship/CrypTen/blob/master/test/multiprocess_test_case.py#L87-L89) to `1`. This will ensure only 1 process is running the unit test code at any given time.

2. Sometimes it is necessary to debug in multiple-processes case to ensure the processes are interacting with each other as expected. This is not likely to be necessary for this task. However, to do this, we adapted the Python debugger to deal with multiple processes. This can be accessed via `crypten.debug.pdb.set_trace()`. 
This will open a command prompt that allows the same debugging functionality as in `pdb`. However, in the CrypTen `pdb` tool, each prompt toggles between the processes (i.e. the first prompt is executed by process 0, the second prompt is executed by process 1, etc.). This can make the CrypTen debugger somewhat annoying to use, however because we adapted this from `pdb`, it is hard to make more user friendly. If you need to use this and want help, please reach out to me.

### PR DESCRIPTION
**Pull request by boyuan12**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

Added _mac attribute initialization, currently, the value is hardcoded to 0. Fixed issue #3 

Issue #3 

## How Has This Been Tested (if it applies)

Added unit test as specified in the issue.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: crypten/mpc/mpc.py
Comment (uhbuhb) [inline crypten/mpc/mpc.py]: nit - blank line above is unnecessary here

File: test/test_mpc.py
Comment (uhbuhb) [inline test/test_mpc.py]: you could add 2 simple tests which unittest the copy_, clone, and shallow_copy methods?

**Review by uhbuhb — COMMENTED** [review]
this PR is made of 3 commits, 1 which is yours and the other 2 are merge commits. 

Better to rebase your branch above the branch you want to merge to so that the only commit will be the commit you made. 

One other nit - as we are here for the learning process - its better practice to separate out code changes to test changes/additions. So I recommend separating your commit into 2 seperate ones. 

In order to make my suggested changes to this PR, you can have a look at the following git commands if you're not yet familiar with them:

```
git rebase
git rebase -i 
git commit --amend
git push -f
```
Good job for putting this up though!

File: test/test_mpc.py
Comment (boyuan12) [reply test/test_mpc.py]: I think when it testing for these three functions([test_copy_](https://github.com/MLH-Fellowship/CrypTen/blob/7d9a070d4ca03ced5772512b22be7078e21cc2ca/test/test_mpc.py#L1520), [test_clone and shallow_copy](https://github.com/MLH-Fellowship/CrypTen/blob/7d9a070d4ca03ced5772512b22be7078e21cc2ca/test/test_mpc.py#L1509)), they have this line: `self._check(encrypted_tensor1, tensor2, "copy_ failed")` at the end, which I believe will run the checks I set in the `_check` method.

Comment (knottb) [reply test/test_mpc.py]: Just some background info:

I noted in the task that these should be here as we want to make sure that adding this functionality to `copy_`, `clone` and `shallow_copy` is sufficient for all other functions to gain this functionality.

I.e. when we call something like `x.add(y)`, the code typically implements this by calling `x.clone().add_(y)`, where `add_()` is an in-place operation. My hypothesis was that because we implement most out-of-place functions by using `clone()` + in-place function, copying the `_mac` attribute in `clone`, etc. would be sufficient.

However, I wanted this unit test to ensure that we weren't missing some functions that may be implemented differently. Because all of our existing unit tests use `_check` to make the proper assertions on the output, adding these assertions to the `_check` function will propagate them to all unit tests. Our existing unit tests should sufficiently cover all functions we care about.

**Review by knottb — APPROVED** [review]

**Comment by knottb** [comment]
Looks good, make the rebase changes suggested by @uhbuhb and this should be good to go!

#shipit

**Comment by mkcode** [comment]
HI Team. Looks like this PR was opened from @boyuan12's fork to the MLH repo. The recommended way to handle this is to create a new feature branch on the MLH-Fellowship fork and the submit the PR to the upstream Facebook organization repo. 

1. Create feature branch locally: `git checkout -b my-new-branch`
2. Push to MLH organization fork: `git push -u mlh $(git branch-name)` (you all have write access and can do this)
3. Submit PR from the MLH fork to the upstream facebook organization master branch.

If you haven't set up your git remotes yet, you can run `git remote add mlh git@github.com:MLH-Fellowship/CrypTen.git` and then use the steps above.

For this PR, I updated the base branch to be a feature branch on the MLH fork. Please merge this PR into that branch, then open a new PR from that feature branch targeting the Facebook organization master branch with a link to this PR so we can preserve the history.

In general, is a good habit to get into to create a new local branch for a new set of changes. You can read more about that here: https://guides.github.com/introduction/flow/

**Comment by smsunarto** [comment]
@mkcode is this good to merge?

**Comment by uhbuhb** [comment]
> @mkcode is this good to merge?

Hey, so the code looks all good, so good job on that. 

A couple of things need to be done before merging this PR:
This pr is to merge boyouan fork master branch into MLH fork add-mac-attribute branch. In reality, what we would like to do is merge from a feature branch on MLH fork to master branch on MLH fork. In order to do that its necessary to:

1. align MLH fork with upstream master
2. create a new branch on MLH fork
3. add the 2 commits on this branch to the new MLH branch
4. Open a PR from the MLH feature branch to the MLH/master branch. 

Can we get this done this week? I'm the exact operations necessary aren't 100% clear I'm happy to help with it, book a pair programming session in my office hours and we can do them together :)

**Comment by boyuan12** [comment]
I think I already add 2 commits to add-mac-attribute branch on MLH fork
    On Tuesday, March 2, 2021, 11:29:27 AM PST, Ori Haberman <notifications@github.com> wrote:  
 
 



@mkcode is this good to merge?


Hey, so the code looks all good, so good job on that.

A couple of things need to be done before merging this PR:
This pr is to merge boyouan fork master branch into MLH fork add-mac-attribute branch. In reality, what we would like to do is merge from a feature branch on MLH fork to master branch on MLH fork. In order to do that its necessary to:
   
   - align MLH fork with upstream master
   - create a new branch on MLH fork
   - add the 2 commits on this branch to the new MLH branch
   - Open a PR from the MLH feature branch to the MLH/master branch.

Can we get this done this week? I'm the exact operations necessary aren't 100% clear I'm happy to help with it, book a pair programming session in my office hours and we can do them together :)

**Comment by uhbuhb** [comment]
> I think I already add 2 commits to add-mac-attribute branch on MLH fork On Tuesday, March 2, 2021, 11:29:27 AM PST, Ori 

So then all you need to do is close this PR and open a new one which will merge add-mac-attribute branch on MLH (this fork) to master branch of this fork.

I can obviously fix it all and put up the PR but were here to learn so I prefer you do it. 

PS - you will probably have to rebase MLH/maser over upstream/master and then rebase your feature branch above MLH/master anyway before opening the PR.

**Pull request merged**
